### PR TITLE
[LTR] Better handling of missing parameters.

### PR DESCRIPTION
**Pull request by afoucret**
### Description 

Currently, params are are replaced by an empty string in query feature extractors when they are missing.

Example: `{ "match": { "field" : "{{param_value}}"}` will be rewritten into `{ "match": { "field" : ""}`

Instead we want the query to always return `0` for the feature (ultimately feature are casted to double, so `0` makes sense). So we would like the query to be rewritten into a `match_none` query.

### Example:

The following config:

```
"learn_to_rank": {
  "feature_extractors": [
    { "query_extractor": { "feature_name": "product_match", "query": {  "match": { "title" : "{{fulltext_search}}" } } } },
    { "query_extractor": { "feature_name": "product_match", "query": {  "term": { "category" : "{{category_id}}" } } } }
  ]
}
```

will be executed as 

```
"learn_to_rank": {
  "feature_extractors": [
    { "query_extractor": { "feature_name": "product_match", "query": {  "match": { "title" : "a super search" } } } },
    { "query_extractor": { "feature_name": "product_match", "query": {  "match_none" : {} } }
  ]
}
```

when called using the following rescorer call:

```
{
  "rescorer" : {
    "model_id" : "my-model",
    "params" : {
      "fulltext_search": "a super search"
    }
  }
}
```

### Implementation detail:

To implement the feature a new `DETECT_MISSING_PARAMS_OPTION` option has been added to compile a mustache script directly into the `MustacheScriptEngine`.

### Example usage:

```
    Map<String, String> SCRIPT_OPTIONS = Map.ofEntries(
        entry(Script.DETECT_MISSING_PARAMS_OPTION, Boolean.TRUE.toString())
    );
    new Script(ScriptType.INLINE, DEFAULT_TEMPLATE_LANG, templateSource, SCRIPT_OPTIONS, Collections.emptyMap());
    String parsedTemplate = scriptService.compile(script, TemplateScript.CONTEXT).newInstance(params).execute();
```

When this options is used an exception is raised when calling `execute` if one of the param present into the template is missing. The option is disabled by default, so the behavior of the `MustacheScriptEngine` is backward compatible.

### CODE REVIEW

File: modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/CustomMustacheFactory.java
Comment (kderusso) [inline modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/CustomMustacheFactory.java:369]: Nitpick: Please create a static final boolean `DEFAULT_DETECT_MISSING_PARAMS` and reference it here for readability.

File: modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/MustacheScriptEngine.java
Comment (kderusso) [inline modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/MustacheScriptEngine.java:120]: Nitpick: Rename method to something like `shouldLogException`

File: modules/lang-mustache/src/test/java/org/elasticsearch/script/mustache/MustacheScriptEngineTests.java
Comment (kderusso) [inline modules/lang-mustache/src/test/java/org/elasticsearch/script/mustache/MustacheScriptEngineTests.java:203]: Do we have tests that will verify we correctly function when detect missing param is false AND params are missing/null?

**Review by kderusso — COMMENTED** [review]
Overall LGTM, some nitpicks.

File: modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/CustomMustacheFactory.java
Comment (afoucret) [reply modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/CustomMustacheFactory.java:369]: Done. Thank you for the feedback.

File: modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/MustacheScriptEngine.java
Comment (afoucret) [reply modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/MustacheScriptEngine.java:120]: :+1

**Comment by benwtrent** [comment]
> Indeed, there are no guarantee that the user will provide default values for each of the template variables present in the query extractors and we will not be able to enforce it.

We should enforce this. Users attempt to create the queries on `PUT` with parameters, if they don't compile to valid query builder objects, we know something failed and they didn't provide correct parameters.

Allowing anything malformed to be `PUT` and users not knowing things are broken until they use it is a bug.

Maybe this can be done in conjunction with your change here that disallows missing parameters.

> Also, my understanding of @joshdevins point is that we want to be able to specify a constant score different than 0.0 when a param is missing. We could add a default_value to the query_extractor definition

Ah, yeah, I didn't think of that. That is probably a way to do provide default values. Instead of a "match_none" they provide a `default_score` or something. I understand.

**Comment by afoucret** [comment]
@benwtrent @joshdevins :

Would you be comfortable if we merge this PR and I implement the following things in a follow-up:
- default value for param at the global level?
- for each query extractor a default score that can be used as a feature value if some missing param prevent us to use the query?

I do not think we can implement the full validation when saving the model before the feature freeze but it should be OK to deliver it anyway since our users will be aware of the tech preview status. 

Also, I think query extractors will be crafted in Eland and saved with the model. I think we can offload the validation to the LTR tooling we will have in Eland and let the training process in charge of validating the query extractors.

**Comment by benwtrent** [comment]
> I do not think we can implement the full validation when saving the model before the feature freeze but it should be OK to deliver it anyway since our users will be aware of the tech preview status.

This is probably ok.

> Also, I think query extractors will be crafted in Eland and saved with the model. I think we can offload the validation to the LTR tooling we will have in Eland and let the training process in charge of validating the query extractors.

As long as this doesn't replace the validation we should be doing in Elasticsearch directly :)

**Review by kderusso — APPROVED** [review]
Thanks for iterating. LGTM pending the open questions in the discussion.

**Comment by afoucret** [comment]
Just finished to prepare a follow-up to this PR: [here](https://github.com/afoucret/elasticsearch/compare/ltr-missing-param-handling...afoucret:elasticsearch:ltr-missing-param-handling-improvements?expand=1).

It is implementing both default param values at the global level and default score for each query extractor:

```
"learn_to_rank": {
  "params": {
    "category_id": 12
  },
  "feature_extractors": [
    { "query_extractor": { "default_score": 2, "feature_name": "product_match", "query": {  "match": { "title" : "{{fulltext_search}}" } } } },
    { "query_extractor": { "feature_name": "product_match", "query": {  "term": { "category" : "{{category_id}}" } } } }
  ]
}
```

I would like to merge this one first and then work on the follow-up. 

@benwtrent @joshdevins Is this OK for you?

**Comment by afoucret** [comment]
@benwtrent Do you prefer that I merge the follow-up PR here or to review both separately?

**Comment by joshdevins** [comment]
Looking good 👍🏻 I'm thinking there might be some usability things we can improve (less verbose maybe) but I'd prefer to save that for later, in a second pass once we have eland, etc. end-to-end working.

**Comment by benwtrent** [comment]
> @benwtrent Do you prefer that I merge the follow-up PR here or to review both separately?

Separate.

The reason I am cautious about this PR is that it messes with the mustache module.

File: modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/CustomReflectionObjectHandler.java
Comment (benwtrent) [inline modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/CustomReflectionObjectHandler.java:61]: @jdconrad ^ this is the method of which I am most suspicious. Does this seem OK to you?

Looking at the mustache plugin code, it does seem ok, it will return a `MissingWrapper` if absolutely nothing is found.

Comment (jdconrad) [reply modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/CustomReflectionObjectHandler.java:61]: This seems fine to me as well as there's no behavioral change as long as `detectMissingParams` defaults to `false`.

File: modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/CustomMustacheFactory.java
Comment (jdconrad) [reply modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/CustomMustacheFactory.java:67]: I think this change is okay as long as we switch this to be created only through a Builder. There should only be way one way to construct these either through standard constructors or through a Builder, but not both.

File: server/src/main/java/org/elasticsearch/script/Script.java
Comment (jdconrad) [inline server/src/main/java/org/elasticsearch/script/Script.java:100]: Not sure this belongs here as it will only apply to Mustache.

**Review by jdconrad — COMMENTED** [review]
Added a few thoughts.

File: modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/CustomMustacheFactory.java
Comment (benwtrent) [reply modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/CustomMustacheFactory.java:67]: At a minimum, what @jdconrad is saying. We need package public ways to construct this factory. We should make the builder package public.

File: modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/CustomMustacheFactory.java
Comment (afoucret) [reply modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/CustomMustacheFactory.java]: I did make it public again but deprecated.

File: modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/CustomMustacheFactory.java
Comment (afoucret) [reply modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/CustomMustacheFactory.java]: Added a javadoc. WDYT?

File: modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/MustacheScriptEngine.java
Comment (afoucret) [reply modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/MustacheScriptEngine.java]: I did update the code so it is now unwrapping only once.

File: modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/CustomReflectionObjectHandler.java
Comment (afoucret) [reply modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/CustomReflectionObjectHandler.java]: I did some change here so we can target only `ValueCode`.
The result is that we can now using conditional with missing parameters without throwing an exception (which is expected from my POV).

Here is the test I did add for it (other tests remain unchanged):

```
// do not throw when using a missing param in the conditional when DETECT_MISSING_PARAMS_OPTION option is set to true
        {
            String source = "{\"match\": { \"field\": \"{{#query_string}}{{.}}{{/query_string}}\" }";
            TemplateScript.Factory compiled = qe.compile(null, source, TemplateScript.CONTEXT, scriptOptions);
            Map<String, Object> params = Map.of();
            assertThat(compiled.newInstance(params).execute(), equalTo("{\"match\": { \"field\": \"\" }"));
        }
```

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ltr/LearnToRankService.java
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ltr/LearnToRankService.java:189]: We should only do a `matchnone` if the script error is specifically because of things missing. Couldn't a `GeneralScriptException` be thrown if things were generally malformed? In the malformed case, we should 100% throw and bubble up the failure.

File: modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/MustacheScriptEngine.java
Comment (afoucret) [reply modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/MustacheScriptEngine.java:129]: The exception is caught into ValueCode before being thrown again wrapped into a MustacheException:

```
  @Override
  public Writer execute(Writer writer, final List<Object> scopes) {
    try {
      final Object object = get(scopes);
      if (object != null) {
        if (object instanceof Function) {
          handleFunction(writer, (Function) object, scopes);
        } else if (object instanceof Callable) {
          return handleCallable(writer, (Callable) object, scopes);
        } else {
          execute(writer, oh.stringify(object));
        }
      }
      return super.execute(writer, scopes);
    } catch (Exception e) {
      throw new MustacheException("Failed to get value for " + name, e, tc);
    }
  }
```

The initial logic was to find an `InvalidParameterException` in the casual chain, so I do not miss one.

An alternative could be:

```java
public boolean shouldLogException(Throwable e) {
    if (e instanceof InvalidParameterException) {
      return false;
    }
    if (e.getCause() != null && e.getCause != e) {
      return shouldLogException(e.getCause());
    }
    return true;
}
```

Comment (benwtrent) [reply modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/MustacheScriptEngine.java:129]: We shouldn't recurse. I am just verifying the logic is ok.

Comment (benwtrent) [reply modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/MustacheScriptEngine.java:129]: I am just cautious as the exception being thrown is from a library we don't own. If we really care about it logging or not, there should be a test that can fail if the behavior within the library changes.

File: server/src/main/java/org/elasticsearch/script/Script.java
Comment (afoucret) [reply server/src/main/java/org/elasticsearch/script/Script.java]: Now a constant of `MustacheScriptEngine`:

I wanted to avoid this initially since it creates a direct dependency to module:mustache for the ml module. But it is required to create this dependency to address this comment: https://github.com/elastic/elasticsearch/pull/102601#discussion_r1410624219

**Review by benwtrent — APPROVED** [review]
:shipit: 

Thank you for iterating on this :D

**Comment by afoucret** [comment]
@benwtrent Thank you for the thoughtful review. 
Let's see if the CI is 🟢  again after I did add the mustache module as an explicit dependency.

I will open the follow-up PR with default params once this one will be merged.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ltr/LearnToRankService.java
Comment (afoucret) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ltr/LearnToRankService.java:192]: @benwtrent Is it possible that the instanceof is return misleading result because of a different classloader for the mustache module?

I have:

```
e.getRootCause() instanceof MustacheInvalidParameterException == false
```

and 

```
e,getRootCause().getClass().getName().equals(MustacheInvalidParameterException.class.getName()) == true
```

Comment (afoucret) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ltr/LearnToRankService.java:192]: Bonus question: what is the cleanest way to write comparison in that case?

Comment (jdconrad) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ltr/LearnToRankService.java:192]: @afoucret If you're genuinely curious you should be able to compare the class loaders from `e.getRootCause()` and `MustacheInvalidParameterException` to confirm this theory. If I remember correctly, we do have to do some magic to get all the class loaders to work correctly with plugins.

**Pull request merged**
